plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java

```diff
@@ -0,0 +1,323 @@
+import com.google.api.services.storage.Storage;
+import com.google.api.services.storage.StorageScopes;
+import com.google.api.services.storage.model.Bucket;
+import com.google.api.services.storage.model.StorageObject;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.blobstore.BlobMetaData;
+import org.elasticsearch.common.blobstore.support.PlainBlobMetaData;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.Callback;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.GeneralSecurityException;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static java.net.HttpURLConnection.HTTP_FORBIDDEN;
+import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
+
+public class InternalGoogleCloudStorageService extends AbstractLifecycleComponent<GoogleCloudStorageService> implements GoogleCloudStorageService {
+
+    /**
+     * Google Cloud Storage batch requests are limited to 1000 operations
+     **/
+    private static final int MAX_BATCHING_REQUESTS = 1_000;
+
+    @Inject
+    public InternalGoogleCloudStorageService(Settings settings) {
+        super(settings);
+    }
+
+    @Override
+    public Client client(Path credentials, String application) {
+        try {
+            logger.debug("loading Google Cloud Storage credentials file from [{}]", credentials);
+            GoogleCredential credential = loadCredentials(credentials);
+
+            logger.debug("building HTTP transport");
+            NetHttpTransport.Builder httpTransport = new NetHttpTransport.Builder();
+            httpTransport.trustCertificates(GoogleUtils.getCertificateTrustStore());
+
+            logger.debug("creating Google Cloud Storage client for application [{}]", application);
+            Storage.Builder storage = new Storage.Builder(httpTransport.build(), JacksonFactory.getDefaultInstance(), credential);
+            storage.setApplicationName(application);
+
+            return new InternalGcsClient(storage.build());
+        } catch (IOException e) {
+            throw new ElasticsearchException("Error when loading Google Cloud Storage credentials file");
+        } catch (GeneralSecurityException e) {
+            throw new ElasticsearchException("Error when loading keystore for Google Cloud Storage client");
+        }
+    }
+
+    @Override
+    protected void doStart() {
+        logger.trace("Google Cloud Storage service started");
+    }
+
+    @Override
+    protected void doStop() {
+        logger.trace("Google Cloud Storage service stopped");
+    }
+
+    @Override
+    protected void doClose() {
+        logger.trace("Google Cloud Storage service closed");
+    }
```

Comment:
Since we are not doing anything in lifecycle methods, why make it into AbstractLifecycleComponent instead of just AbstractComponent in the first place?